x/reporter/reporter.go

```diff
@@ -0,0 +1,211 @@
+	// Observations
+	Time       time.Time   `json:"time"`
+	DurationMs int64       `json:"durationMs"`
+	Error      interface{} `json:"error"`
+}
+
+type Report interface {
+	IsSuccess() bool
+	CanMarshal() bool
+}
+
+// ConnectivityReport implements the Report interface
+func (r ConnectivityReport) IsSuccess() bool {
+	if r.Error == nil {
+		return true
+	} else {
+		return false
+	}
+}
+
+// Makes sure the Report type can be marshalled into JSON
+func (r ConnectivityReport) CanMarshal() bool {
+	_, err := json.Marshal(r)
+	if err != nil {
+		log.Printf("Error encoding JSON: %s\n", err)
+		return false
+	} else {
+		return true
+	}
+}
+
+type Collector interface {
+	Collect(Report) error
+}
+
+type RemoteCollector struct {
+	collectorEndpoint *url.URL
+}
+
+type SamplingCollector struct {
+	collector       Collector
+	successFraction float64
+	failureFraction float64
+}
+
+type CollectorTarget struct {
+	collector Collector
+	priority  int
+	maxRetry  int
+}
+
+// TODO: implement a rotating collector
+type RotatingCollector struct {
+	collectors    []CollectorTarget
+	stopOnSuccess bool
+}
+
+// SortByPriority sorts the collectors based on their priority in ascending order
+func (rc *RotatingCollector) SortByPriority() {
+	sort.Slice(rc.collectors, func(i, j int) bool {
+		return rc.collectors[i].priority < rc.collectors[j].priority
+	})
+}
+
+func (c *RotatingCollector) Collect(report Report) error {
+	// sort collectors in RotatingCollector by priority
+	// into a new slice
+	c.SortByPriority()
+	for _, target := range c.collectors {
+		for i := 0; i < target.maxRetry; i++ {
+			err := target.collector.Collect(report)
+			if err != nil {
+				if err, ok := err.(StatusErr); ok {
+					switch {
+					case err.StatusCode == 500:
+						// skip retrying
+						break
+					case err.StatusCode == 408:
+						// wait for 1 second before retry
+						time.Sleep(time.Duration(1000 * time.Millisecond))
+					default:
+						break
+					}
+				} else {
+					return err
+				}
+			} else {
+				fmt.Println("Report sent")
+				if c.stopOnSuccess {
+					return nil
+				}
+				break
+			}
+		}
+	}
+	return nil
+}
+
+type StatusErr struct {
+	StatusCode int
+	Message    string
+}
+
+func (e StatusErr) Error() string {
+	return e.Message
+}
+
+func (c *RemoteCollector) Collect(report Report) error {
+	jsonData, err := json.Marshal(report)
+	var statusCode int
+	if err != nil {
+		log.Printf("Error encoding JSON: %s\n", err)
+		return err
+	}
+	statusCode, err = sendReport(jsonData, c.collectorEndpoint)
+	if err != nil {
+		log.Printf("Send report failed: %v", err)
+		return err
+	} else if statusCode > 400 {
+		return StatusErr{
+			StatusCode: statusCode,
+			Message:    fmt.Sprintf("HTTP request failed with status code %d", statusCode),
+		}
+	} else {
+		fmt.Println("Report sent")
+		return nil
+	}
+}
+
+func (c *SamplingCollector) Collect(report Report) error {
+	var samplingRate float64
+	if report.IsSuccess() {
```

Comment:
```suggestion
	hs, ok := report.(HasSuccess)
	if !ok {
	    return error.New("report for SamplingCollector must implement HasSuccess")
	}
	if hs.IsSuccess() {
```

Reply:
The error should be a `BadRequest`, so the retrier knows it shouldn't retry.

Reply:
`BadRequestErr` is now defined for all HTTP error codes >= 400. I am passing the actual error code inside  `BadRequestErr`